backend/x86_binary_emitter.ml

```diff
@@ -0,0 +1,1764 @@
+end)
+
+module StringMap = Map.Make (struct
+    type t = string
+
+    let compare = String.compare
+  end)
+
+let print_old_arg ppf = function
+  | Imm _ -> Format.fprintf ppf "Imm"
+  | Reg8L _ -> Format.fprintf ppf "Reg8L"
+  | Reg8H _ -> Format.fprintf ppf "Reg8H"
+  | Reg16 _ -> Format.fprintf ppf "Reg16"
+  | Reg32 _ -> Format.fprintf ppf "Reg32"
+  | Reg64 _ -> Format.fprintf ppf "Reg64"
+  | Regf _ -> Format.fprintf ppf "Regf"
+  | Mem _ -> Format.fprintf ppf "Mem"
+  | Mem64_RIP _ -> Format.fprintf ppf "Mem64_RIP"
+  | Sym _ -> Format.fprintf ppf "Sym"
+
+(*
+TODO:
+
+If a or-pattern contains both "Reg64 ... | Reg32 ... ", it means that
+we didn't discriminate between 32 bit and 64 bit modes for that
+instruction. It also means that using this instruction on a 32-bit
+register in 64 bit mode will not generate the 32-bit version of the
+instruction, but the 64-bit version...
+
+*)
+
+module Relocation = struct
+  module Kind = struct
+    type t =
+      (* 32 bits offset usually in data section *)
+      | REL32 of string * int64
+      | DIR32 of string * int64
+      | DIR64 of string * int64
+  end
+
+  type t = { offset_from_section_beginning : int; kind : Kind.t }
+end
+
+type symbol = {
+  sy_name : string;
+  mutable sy_type : string option;
+  mutable sy_size : int option;
+  mutable sy_global : bool;
+  mutable sy_sec : section;
+  mutable sy_pos : int option;
+  mutable sy_num : int option; (* position in .symtab *)
+}
+
+type buffer = {
+  sec : section;
+  buf : Buffer.t;
+  mutable labels : symbol StringMap.t;
+  mutable patches : (int * data_size * int64) list;
+  mutable relocations : Relocation.t list;
+}
+
+type local_reloc =
+  | RelocCall of string
+  | RelocShortJump of string * int (* loc *)
+  | RelocLongJump of string
+  | RelocConstant of constant * data_size
+
+type result =
+  | Rint of int64
+  | Rabs of string * int64 (* absolute label + offset *)
+  | Rrel of string * int64
+
+(* relative label + offset *)
+
+(*
+let string_of_result = function
+  Rint n -> Printf.sprintf "Rint %Ld" n
+  | Rabs (s, n) -> Printf.sprintf "Rabs (%S, %Ld)" s n
+  | Rrel (s, n) -> Printf.sprintf "Rrel (%S, %Ld)" s n
+*)
+
+let get_symbol b s =
+  try StringMap.find s b.labels
+  with Not_found ->
+    let sy =
+      {
+        sy_name = s;
+        sy_type = None;
+        sy_size = None;
+        sy_pos = None;
+        sy_global = false;
+        sy_num = None;
+        sy_sec = b.sec;
+      }
+    in
+    b.labels <- StringMap.add s sy b.labels;
+    sy
+
+let buf_int8 b i = Buffer.add_char b.buf (char_of_int (i land 0xff))
+
+let buf_int8L b iL = buf_int8 b (Int64.to_int iL)
+
+let buf_int16L b iL =
+  buf_int8L b iL;
+  buf_int8L b (Int64.shift_right iL 8)
+
+let buf_int32L b iL =
+  buf_int16L b iL;
+  buf_int16L b (Int64.shift_right iL 16)
+
+let buf_int64L b iL =
+  buf_int32L b iL;
+  buf_int32L b (Int64.shift_right iL 32)
+
+let str_int8L s pos v = Bytes.set s pos (char_of_int (Int64.to_int v land 0xff))
```

Comment:
(Same remark with function from [this section](https://github.com/ocaml-flambda/flambda-backend/blob/main/ocaml/stdlib/bytes.mli#L474).)